File: src/import-plan.js
import { readFileSync } from 'node:fs';
import { basename } from 'node:path';
import { projects, plans, units, tasks } from './repo.js';

// Parse a Claude Code plan markdown file into Clawket entities.
// Rules (deterministic, no LLM):
//   - First `# <title>` line is the Plan title.
//   - `## Unit N: <title>` (or `## Unit: <title>`) sections become Unit rows.
//     If no explicit Unit heading found, the whole plan becomes a single "Unit 1".
//   - Within a Unit, `### <title>` headings become Tasks.
//     Or, if the Unit contains a numbered list ("1. foo", "2. bar") at the top level,
//     those become Tasks instead.
//   - The text content between headings becomes the body of the enclosing entity.

export function parsePlanMarkdown(md) {
  const lines = md.split('\n');
  const plan = { title: null, description: '', units: [] };

  // Find plan title
  const titleMatch = lines.find(l => /^#\s+/.test(l));
  if (titleMatch) plan.title = titleMatch.replace(/^#\s+/, '').trim();

  // Find unit sections by looking for ## Unit N: or ## Unit: headings
  const unitHeadingRe = /^##\s+Unit\s*(\d+)?\s*[:.]?\s*(.*)$/i;

  // Locate unit headings
  const unitMarkers = [];
  for (let i = 0; i < lines.length; i++) {
    const m = lines[i].match(unitHeadingRe);
    if (m) {
      unitMarkers.push({
        lineIdx: i,
        idx: unitMarkers.length,
        title: (m[2] && m[2].trim()) || `Unit ${m[1] || unitMarkers.length + 1}`,
      });
    }
  }

  if (unitMarkers.length === 0) {
    // No explicit Unit section: synthesize a single Unit containing all content
    plan.units.push({
      idx: 0,
      title: 'Unit 1',
      goal: null,
      body: lines.join('\n').trim(),
      tasks: extractTasksFromBody(lines, 0, lines.length),
    });
  } else {
    // Description = content from start to first unit
    plan.description = lines.slice(0, unitMarkers[0].lineIdx).join('\n').trim();

    for (let p = 0; p < unitMarkers.length; p++) {
      const start = unitMarkers[p].lineIdx;
      const end = p + 1 < unitMarkers.length ? unitMarkers[p + 1].lineIdx : lines.length;
      plan.units.push({
        idx: p,
        title: unitMarkers[p].title,
        goal: null,
        body: lines.slice(start, end).join('\n').trim(),
        tasks: extractTasksFromBody(lines, start + 1, end),
      });
    }
  }

  return plan;
}

function extractTasksFromBody(lines, start, end) {
  // Prefer ### subheadings as tasks
  const h3Re = /^###\s+(.+)$/;
  const h3Indices = [];
  for (let i = start; i < end; i++) {
    const m = lines[i].match(h3Re);
    if (m) h3Indices.push({ lineIdx: i, title: m[1].trim() });
  }

  if (h3Indices.length > 0) {
    const out = [];
    for (let j = 0; j < h3Indices.length; j++) {
      const s = h3Indices[j].lineIdx;
      const e = j + 1 < h3Indices.length ? h3Indices[j + 1].lineIdx : end;
      out.push({
        idx: j,
        title: h3Indices[j].title,
        body: lines.slice(s + 1, e).join('\n').trim(),
      });
    }
    return out;
  }

  // Fallback: look for numbered list items at paragraph starts
  const numberedRe = /^\s*\d+\.\s+\*?\*?([^*]+)\*?\*?(.*)$/;
  const out = [];
  for (let i = start; i < end; i++) {
    const m = lines[i].match(numberedRe);
    if (m) {
      out.push({
        idx: out.length,
        title: (m[1] || '').trim(),
        body: ((m[2] || '') + '\n' + gatherContinuation(lines, i + 1, end)).trim(),
      });
    }
  }
  return out;
}

function gatherContinuation(lines, from, end) {
  // Collect lines until next list item or heading
  const out = [];
  for (let i = from; i < end; i++) {
    if (/^##?#?\s+/.test(lines[i]) || /^\s*\d+\.\s+/.test(lines[i])) break;
    out.push(lines[i]);
  }
  return out.join('\n');
}

export function importPlanFile(filePath, { projectName = null, cwd = null, source = 'import', dryRun = false } = {}) {
  const md = readFileSync(filePath, 'utf8');
  const parsed = parsePlanMarkdown(md);
  if (!parsed.title) throw new Error('no plan title (first # heading) found');

  // Resolve project
  let project;
  if (projectName) {
    project = projects.getByName(projectName);
    if (!project) {
      project = projects.create({ name: projectName, cwd: cwd || process.cwd() });
    }
  } else {
    project = projects.getByCwd(cwd || process.cwd());
    if (!project) {
      const fallback = basename(filePath, '.md').slice(0, 40);
      project = projects.create({ name: fallback, cwd: cwd || process.cwd() });
    }
  }

  if (dryRun) {
    return {
      dryRun: true,
      project: project,
      plan_title: parsed.title,
      unit_count: parsed.units.length,
      task_count: parsed.units.reduce((n, u) => n + u.tasks.length, 0),
      units: parsed.units.map(u => ({ title: u.title, tasks: u.tasks.map(t => t.title) })),
    };
  }

  // Create plan
  const plan = plans.create({
    project_id: project.id,
    title: parsed.title,
    description: parsed.description || null,
    source,
    source_path: filePath,
  });

  // Create units + tasks
  const createdUnits = [];
  for (const u of parsed.units) {
    const unitRow = units.create({
      plan_id: plan.id,
      title: u.title,
      goal: null,
      idx: u.idx,
      approval_required: false,
    });
    const createdTasks = [];
    for (const t of u.tasks) {
      const taskRow = tasks.create({
        unit_id: unitRow.id,
        title: t.title,
        body: t.body,
        idx: t.idx,
      });
      createdTasks.push(taskRow);
    }
    createdUnits.push({ unit: unitRow, tasks: createdTasks });
  }

  return {
    project,
    plan,
    units: createdUnits,
    summary: {
      unit_count: createdUnits.length,
      task_count: createdUnits.reduce((n, u) => n + u.tasks.length, 0),
    },
  };
}

File: rust/src/import_plan.rs
use anyhow::{bail, Result};
use once_cell::sync::Lazy;
use regex::Regex;
use std::path::Path;

use crate::models::{Plan, Project, Task, Unit};
use crate::repo::{plans, projects, tasks, units};
use rusqlite::Connection;

#[derive(Debug, Default, Clone)]
pub struct ParsedTask {
    pub idx: i64,
    pub title: String,
    pub body: String,
}

#[derive(Debug, Default, Clone)]
pub struct ParsedUnit {
    pub idx: i64,
    pub title: String,
    pub body: String,
    pub tasks: Vec<ParsedTask>,
}

#[derive(Debug, Default, Clone)]
pub struct ParsedPlan {
    pub title: Option<String>,
    pub description: String,
    pub units: Vec<ParsedUnit>,
}

static UNIT_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"(?i)^##\s+Unit\s*(\d+)?\s*[:.]?\s*(.*)$").unwrap()
});
static H3_RE: Lazy<Regex> = Lazy::new(|| Regex::new(r"^###\s+(.+)$").unwrap());
static NUMBERED_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"^\s*\d+\.\s+\*?\*?([^*]+)\*?\*?(.*)$").unwrap());
static HEADING_ANY_RE: Lazy<Regex> = Lazy::new(|| Regex::new(r"^##?#?\s+").unwrap());
static NUMBERED_START_RE: Lazy<Regex> = Lazy::new(|| Regex::new(r"^\s*\d+\.\s+").unwrap());
static TITLE_RE: Lazy<Regex> = Lazy::new(|| Regex::new(r"^#\s+").unwrap());

pub fn parse_plan_markdown(md: &str) -> ParsedPlan {
    let lines: Vec<&str> = md.split('\n').collect();
    let mut plan = ParsedPlan::default();

    if let Some(t) = lines.iter().find(|l| TITLE_RE.is_match(l)) {
        let stripped = TITLE_RE.replace(t, "").trim().to_string();
        plan.title = Some(stripped);
    }

    let mut unit_markers: Vec<(usize, String)> = Vec::new();
    for (i, line) in lines.iter().enumerate() {
        if let Some(caps) = UNIT_RE.captures(line) {
            let num = caps.get(1).map(|m| m.as_str().to_string());
            let name = caps
                .get(2)
                .map(|m| m.as_str().trim().to_string())
                .filter(|s| !s.is_empty());
            let title = name.unwrap_or_else(|| {
                format!("Unit {}", num.unwrap_or_else(|| (unit_markers.len() + 1).to_string()))
            });
            unit_markers.push((i, title));
        }
    }

    if unit_markers.is_empty() {
        plan.units.push(ParsedUnit {
            idx: 0,
            title: "Unit 1".to_string(),
            body: lines.join("\n").trim().to_string(),
            tasks: extract_tasks_from_body(&lines, 0, lines.len()),
        });
    } else {
        plan.description = lines[..unit_markers[0].0].join("\n").trim().to_string();
        for (p, (line_idx, title)) in unit_markers.iter().enumerate() {
            let end = unit_markers
                .get(p + 1)
                .map(|(i, _)| *i)
                .unwrap_or(lines.len());
            plan.units.push(ParsedUnit {
                idx: p as i64,
                title: title.clone(),
                body: lines[*line_idx..end].join("\n").trim().to_string(),
                tasks: extract_tasks_from_body(&lines, *line_idx + 1, end),
            });
        }
    }

    plan
}

fn extract_tasks_from_body(lines: &[&str], start: usize, end: usize) -> Vec<ParsedTask> {
    let mut h3: Vec<(usize, String)> = Vec::new();
    for i in start..end {
        if let Some(caps) = H3_RE.captures(lines[i]) {
            h3.push((i, caps[1].trim().to_string()));
        }
    }
    if !h3.is_empty() {
        let mut out = Vec::new();
        for (j, (s, title)) in h3.iter().enumerate() {
            let e = h3.get(j + 1).map(|(i, _)| *i).unwrap_or(end);
            out.push(ParsedTask {
                idx: j as i64,
                title: title.clone(),
                body: lines[s + 1..e].join("\n").trim().to_string(),
            });
        }
        return out;
    }

    let mut out: Vec<ParsedTask> = Vec::new();
    for i in start..end {
        if let Some(caps) = NUMBERED_RE.captures(lines[i]) {
            let title = caps
                .get(1)
                .map(|m| m.as_str().trim().to_string())
                .unwrap_or_default();
            let tail = caps
                .get(2)
                .map(|m| m.as_str().to_string())
                .unwrap_or_default();
            let body = format!("{}\n{}", tail, gather_continuation(lines, i + 1, end))
                .trim()
                .to_string();
            out.push(ParsedTask {
                idx: out.len() as i64,
                title,
                body,
            });
        }
    }
    out
}

fn gather_continuation(lines: &[&str], from: usize, end: usize) -> String {
    let mut out: Vec<&str> = Vec::new();
    for i in from..end {
        if HEADING_ANY_RE.is_match(lines[i]) || NUMBERED_START_RE.is_match(lines[i]) {
            break;
        }
        out.push(lines[i]);
    }
    out.join("\n")
}

pub struct ImportOptions<'a> {
    pub project_name: Option<&'a str>,
    pub cwd: Option<&'a str>,
    pub source: &'a str,
    pub dry_run: bool,
}

#[derive(serde::Serialize)]
pub struct DryRunUnit {
    pub title: String,
    pub tasks: Vec<String>,
}

#[derive(serde::Serialize)]
#[serde(untagged)]
pub enum ImportResult {
    DryRun {
        dry_run: bool,
        project: Project,
        plan_title: String,
        unit_count: usize,
        task_count: usize,
        units: Vec<DryRunUnit>,
    },
    Created {
        project: Project,
        plan: Plan,
        units: Vec<CreatedUnit>,
        summary: Summary,
    },
}

#[derive(serde::Serialize)]
pub struct CreatedUnit {
    pub unit: Unit,
    pub tasks: Vec<Task>,
}

#[derive(serde::Serialize)]
pub struct Summary {
    pub unit_count: usize,
    pub task_count: usize,
}

pub fn import_plan_file(
    conn: &mut Connection,
    file_path: &str,
    opts: ImportOptions<'_>,
) -> Result<ImportResult> {
    let md = std::fs::read_to_string(file_path)?;
    let parsed = parse_plan_markdown(&md);
    let Some(title) = parsed.title.clone() else {
        bail!("no plan title (first # heading) found");
    };

    let effective_cwd = opts
        .cwd
        .map(String::from)
        .unwrap_or_else(|| {
            std::env::current_dir()
                .map(|p| p.to_string_lossy().into_owned())
                .unwrap_or_default()
        });

    let project = if let Some(name) = opts.project_name {
        match projects::get_by_name(conn, name)? {
            Some(p) => p,
            None => projects::create(
                conn,
                projects::CreateInput {
                    name,
                    description: None,
                    cwd: Some(&effective_cwd),
                    key: None,
                },
            )?
            .ok_or_else(|| anyhow::anyhow!("project create failed"))?,
        }
    } else {
        match projects::get_by_cwd(conn, &effective_cwd, false)? {
            Some(p) => p,
            None => {
                let fallback = Path::new(file_path)
                    .file_stem()
                    .and_then(|s| s.to_str())
                    .unwrap_or("plan")
                    .chars()
                    .take(40)
                    .collect::<String>();
                projects::create(
                    conn,
                    projects::CreateInput {
                        name: &fallback,
                        description: None,
                        cwd: Some(&effective_cwd),
                        key: None,
                    },
                )?
                .ok_or_else(|| anyhow::anyhow!("project create failed"))?
            }
        }
    };

    if opts.dry_run {
        let unit_count = parsed.units.len();
        let task_count: usize = parsed.units.iter().map(|u| u.tasks.len()).sum();
        let units_out: Vec<DryRunUnit> = parsed
            .units
            .iter()
            .map(|u| DryRunUnit {
                title: u.title.clone(),
                tasks: u.tasks.iter().map(|t| t.title.clone()).collect(),
            })
            .collect();
        return Ok(ImportResult::DryRun {
            dry_run: true,
            project,
            plan_title: title,
            unit_count,
            task_count,
            units: units_out,
        });
    }

    let description = if parsed.description.is_empty() {
        None
    } else {
        Some(parsed.description.as_str())
    };
    let plan = plans::create(
        conn,
        plans::CreateInput {
            project_id: &project.id,
            title: &title,
            description,
            source: Some(opts.source),
            source_path: Some(file_path),
        },
    )?
    .ok_or_else(|| anyhow::anyhow!("plan create failed"))?;

    let mut created_units: Vec<CreatedUnit> = Vec::new();
    for pu in &parsed.units {
        let unit = units::create(
            conn,
            units::CreateInput {
                plan_id: &plan.id,
                title: &pu.title,
                goal: None,
                idx: Some(pu.idx),
                approval_required: false,
                execution_mode: None,
            },
        )?
        .ok_or_else(|| anyhow::anyhow!("unit create failed"))?;
        let mut created_tasks = Vec::new();
        for pt in &pu.tasks {
            let task = tasks::create(
                conn,
                tasks::CreateInput {
                    unit_id: &unit.id,
                    title: &pt.title,
                    body: Some(&pt.body),
                    assignee: None,
                    idx: Some(pt.idx),
                    depends_on: Vec::new(),
                    parent_task_id: None,
                    priority: None,
                    complexity: None,
                    estimated_edits: None,
                    cycle_id: None,
                    reporter: None,
                    type_: None,
                },
            )?
            .ok_or_else(|| anyhow::anyhow!("task create failed"))?;
            created_tasks.push(task);
        }
        created_units.push(CreatedUnit {
            unit,
            tasks: created_tasks,
        });
    }

    let unit_count = created_units.len();
    let task_count: usize = created_units.iter().map(|u| u.tasks.len()).sum();
    Ok(ImportResult::Created {
        project,
        plan,
        units: created_units,
        summary: Summary {
            unit_count,
            task_count,
        },
    })
}
